README.md
# 📝 logit

[![Go Doc](_icons/godoc.svg)](https://pkg.go.dev/github.com/creky/logit)
[![License](_icons/license.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)
[![Build](_icons/build.svg)](_icons/build.svg)
[![Coverage](_icons/coverage.svg)](_icons/coverage.svg)

**logit** 是一个基于级别控制的高性能纯结构化日志库，可以应用于所有的 [GoLang](https://golang.org) 应用程序中。

[Read me in English](./README.en.md)

### 🥇 功能特性

* 独特的日志模块设计，使用 appender 和 writer 装载特定的模块，实现扩展功能。
* 支持日志级别控制，一共有五个日志级别，分别是 debug，info，warn，error，print 和 off。
* 支持键值对形式的结构化日志记录，同时对格式化操作也有支持。
* 支持以 Text/Json 形式输出日志信息，方便对日志进行解析。
* 支持异步回写日志，提供高性能缓冲写出器模块，减少 IO 的访问次数。
* 提供调优使用的全局配置，对一些高级配置更贴合实际业务的需求。
* 分级别追加日志数据，分级别写出日志数据，推荐将 error 级别的日志单独处理和存储。
* 加入 Context 机制，更优雅地使用日志，并支持业务域划分。
* 支持拦截器模式，可以从 context 注入外部常量或变量，简化日志输出流程。
* 支持错误监控，可以很方便地进行错误统计和告警。
* 支持日志按大小自动分割，并支持按照时间和数量自动清理。
* 支持多种配置文件序列化成 option，比如 json/yaml/toml/bson，然后创建日志记录器。

> 更多 logit 生态的东西请查看 [GitHub](https://github.com/go-logit) / [码云](https://gitee.com/go-logit)

_历史版本的特性请查看 [HISTORY.md](./HISTORY.md)。未来版本的新特性和计划请查看 [FUTURE.md](./FUTURE.md)。_

### 🚀 安装方式

```bash
$ go get -u github.com/creky/logit
```

### 📖 参考案例

```go
package main

import (
	"context"
	"io"
	"os"

	"github.com/creky/logit"
	"github.com/creky/logit/support/global"
)

func main() {
	// Create a new logger for use.
	// Default level is debug, so all logs will be logged.
	// Invoke Close() isn't necessary in all situations.
	// If logger's writer has buffer or something like that, it's better to invoke Close() for syncing buffer or something else.
	logger := logit.NewLogger()
	//defer logger.Close()

	// Then, you can log anything you want.
	// Remember, logs will be ignored if their level is smaller than logger's level.
	// Log() will do some finishing work, so this invocation is necessary.
	logger.Debug("This is a debug message").Log()
	logger.Info("This is an info message").Log()
	logger.Warn("This is a warn message").Log()
	logger.Error("This is an error message").Log()
	logger.Error("This is a %s message, with format", "error").Log() // Format with params.

	// As you know, we provide some levels: debug, info, warn, error, off.
	// The lowest is debug and the highest is off.
	// If you want to change the level of your logger, do it at creating.
	logger = logit.NewLogger(logit.Options().WithWarnLevel())
	logger.Debug("This is a debug message, but ignored").Log()
	logger.Info("This is an info message, but ignored").Log()
	logger.Warn("This is a warn message, not ignored").Log()
	logger.Error("This is an error message, not ignored").Log()

	// Also, we provide some "old school" log method :)
	// (Don't mistake~ I love old school~)
	logger.Printf("This is a log %s, and it's for compatibility", "printed")
	logger.Print("This is a log printed, and it's for compatibility", 123)
	logger.Println("This is a log printed, and it's for compatibility", 666)

	// If you want to log with some fields, try this:
	user := struct {
		ID   int64  `json:"id"`
		Name string `json:"name"`
		Age  int    `json:"age"`
	}{
		ID:   666,
		Name: "FishGoddess",
		Age:  3,
	}
	logger.Warn("This is a structured message").Any("user", user).Json("userJson", user).Log()
	logger.Error("This is a structured message").Error("err", io.EOF).Int("trace", 123).Log()

	// You may notice logit.Options() which returns an options list.
	// Here is some of them:
	options := logit.Options()
	options.WithCaller()                          // Let logs carry caller information.
	options.WithLevelKey("lvl")                   // Change logger's level key to "lvl".
	options.WithWriter(os.Stderr)                 // Change logger's writer to os.Stderr without buffer or batch.
	options.WithBufferWriter(os.Stderr)           // Change logger's writer to os.Stderr with buffer.
	options.WithBatchWriter(os.Stderr)            // Change logger's writer to os.Stderr with batch.
	options.WithErrorWriter(os.Stderr)            // Change logger's error writer to os.Stderr without buffer or batch.
	options.WithTimeFormat("2006-01-02 15:04:05") // Change the format of time (Only the log's time will apply it).

	// You can bind context with logger and use it as long as you can get the context.
	ctx := logit.NewContext(context.Background(), logger)
	logger = logit.FromContext(ctx)
	logger.Info("Logger from context").Log()

	// You can initialize the global logger if you don't want to use an independent logger.
	// WithCallerDepth will set the depth of caller, and default is core.CallerDepth.
	// Functions in global logger are wrapped so depth of caller should be increased 1.
	// You can specify your depth if you wrap again or have something else reasons.
	logger = logit.NewLogger(options.WithCallerDepth(global.CallerDepth + 1))
	logit.SetGlobal(logger)
	logit.Info("Info from logit").Log()

	// We don't recommend you to call logit.SetGlobal unless you really need to call.
	// Instead, we recommend you to call logger.SetToGlobal to set one logger to global if you need.
	logger.SetToGlobal()
	logit.Println("Println from logit")
}
```

_更多使用案例请查看 [_examples](./_examples) 目录。_

### 🔥 性能测试

```bash
$ make bench
$ make benchfile
```

| 测试（输出到内存） | 单位时间内运行次数 (越大越好) | 每个操作消耗时间 (越小越好)       | B/op (越小越好)                     | allocs/op (越小越好)              |
|-----------|------------------|-----------------------|---------------------------------|-------------------------------|
| **logit** | **707851**       | **&nbsp; 1704 ns/op** | **&nbsp; &nbsp; &nbsp; 0 B/op** | **&nbsp; &nbsp; 0 allocs/op** |
| zerolog   | 706714           | &nbsp; 1585 ns/op     | &nbsp; &nbsp; &nbsp; 0 B/op     | &nbsp; &nbsp; 0 allocs/op     |
| zap       | 389608           | &nbsp; 4688 ns/op     | &nbsp; 865 B/op                 | &nbsp; &nbsp; 8 allocs/op     |
| logrus    | &nbsp; 69789     | 17142 ns/op           | 8885 B/op                       | 136 allocs/op                 |

| 测试（输出到文件）     | 单位时间内运行次数 (越大越好) | 每个操作消耗时间 (越小越好)       | B/op (越小越好)                     | allocs/op (越小越好)              |
|---------------|------------------|-----------------------|---------------------------------|-------------------------------|
| **logit**     | **636033**       | **&nbsp; 1822 ns/op** | **&nbsp; &nbsp; &nbsp; 0 B/op** | **&nbsp; &nbsp; 0 allocs/op** |
| **logit-不缓冲** | **354542**       | **&nbsp; 3502 ns/op** | **&nbsp; &nbsp; &nbsp; 0 B/op** | **&nbsp; &nbsp; 0 allocs/op** |

HISTORY.md
## ✒ 历史版本的特性介绍 (Features in old versions)

### v0.5.5-alpha

> 此版本发布于 2022-09-02

* 大量完善单元测试
* 大量优化代码设计
* 修复 WithPID 在默认开启时不会添加 pid 的问题

### v0.5.4-alpha

> 此版本发布于 2022-09-01

* 大幅度优化 API 和代码（再忙也要学会优雅哦，哈哈哈）
* 增加配置项解析功能，提供配置文件支持
* 加入日志存活天数的特性
* 加入日志存活个数的特性
* 按时间、文件大小自动切割日志文件，并支持过期清理机制

### v0.5.3-alpha

> 此版本发布于 2022-08-28

* 调整包结构，优化代码
* 增加配置文件解析机制

### v0.5.2-alpha

> 此版本发布于 2022-08-27

* 增加错误监控回调函数

### v0.5.1-alpha

> 此版本发布于 2022-08-24

* 提取时间函数，方便做时间查询优化
* 优化 Json 方法的 Appender 实现

### v0.5.0-alpha

> 此版本发布于 2022-08-22

* 替换 End() 方法，使用 Log() 代替，代码可读性更高
* 代码优化和重构，包括性能和逻辑上的优化，代码质量上的重构
* 支持增加调用堆栈或者函数名的信息到日志
* 增加批量写入的 Writer，区别于缓冲写入的 Writer，这个批量是按照数量进行缓冲的

### v0.4.22

> 此版本发布于 2022-03-19

* 前两个测试版本的正式发布版本

### v0.4.21-alpha

> 此版本发布于 2022-02-27

* 废弃 LoggerMaker
* 加入全局日志方法

### v0.4.20-alpha

> 此版本发布于 2022-02-23

* 加入日志拦截器，支持从 context 注入键值对
* 优化 Logger 字段结构，减少多余引用变量
* 更改 LoggerMaker 为 LoggerCreator，并废弃 LoggerMaker
* 更改 MarshalJson 为 MarshalToJson，并废弃 MarshalJson

### v0.4.19

> 此版本发布于 2022-02-13

* 给 log 加入 Json 方法

### v0.4.18

> 此版本发布于 2022-02-08

* 修复 Caller Depth 不生效的问题

### v0.4.17

> 此版本发布于 2022-02-08

* 修复 Version 错误的问题
* 修复 Printf、Print、Println 方法下 level 错误的问题
* 加入 WithPid 方法
* 完善 PrintLevel 相关的逻辑

### v0.4.16

> 此版本发布于 2022-02-02

* 庆祝下仓库转移成功！！！

### v0.4.15

> 此版本发布于 2022-02-01

* 祝大家新春快乐 :)

### v0.4.14-alpha

> 此版本发布于 2022-01-31

* 兼容 log 包部分方法，比如 Printf

### v0.4.13-alpha

> 此版本发布于 2022-01-29

* Log 增加 Caller 方法，方便某些场景下获取调用信息

### v0.4.12-alpha

> 此版本发布于 2022-01-29

* depth 放开，可以支持用户包裹

### v0.4.11

> 此版本发布于 2021-11-29

* 修改 LoggerMaker 接口 params 为可变参数

### v0.4.10

> 此版本发布于 2021-11-01

* 增加 LoggerMaker 扩展功能

### v0.4.9-alpha

> 此版本发布于 2021-10-27

* 修复 TextAppender 的 AppendErrors 和 AppendStringers 追加问题
* 完善单元测试

### v0.4.8-alpha

> 此版本发布于 2021-10-10

* 修改 sync.Mutex 的使用方式
* 完善单元测试

### v0.4.7-alpha

> 此版本发布于 2021-09-27

* 加入 Context 机制，更优雅地使用日志，并支持业务域划分

### v0.4.6-alpha

> 此版本发布于 2021-09-08

* 调整转义字符显示样式

### v0.4.5-alpha

> 此版本发布于 2021-09-06

* 修复 Appender 接口实现类检验失效问题
* 调整 textAppender 的字段分隔符为 |
* 去除 signal 机制的 TODO

### v0.4.4-alpha

> 此版本发布于 2021-08-10

* 修复 WithWriter 的顺序问题

### v0.4.3-alpha

> 此版本发布于 2021-08-10

* 优化部分 log 设计
* 增加多级别追加器机制
* 增加多级别写出器机制
* 完善文档注释和单元测试

### v0.4.2-alpha

> 此版本发布于 2021-08-01

* 修复部分问题
* 完善文档注释

### v0.4.1-alpha

> 此版本发布于 2021-07-15

* 改造 API 使用体验，加入结构化日志功能
* Options 机制引入，支持多种创建选项
* 改造 Encoder 为 Appender，直接追加字节数据
* 提供高级调优配置能力，贴合业务进行设置

### v0.4.0-alpha

> 此版本发布于 2021-06-08

* 大幅度重构设计，精简功能，后续继续完善原有功能
* 重构 Encoder，加入缓冲区对象池
* 引入新的日志输出器，并且带有缓冲和异步写功能

### v0.3.3

> 此版本发布于 2021-01-12

* 去除了 DebugF 一类格式化的 API，统一使用 Debug 一类 API

### v0.3.2

> 此版本发布于 2020-12-24

* 废弃了 files 包
* 完善文档，第一个 v0.3.x 正式版
* 最后，祝大家平安夜、圣诞节快乐！

### v0.3.1-alpha

> 此版本发布于 2020-12-13

* 日志文件自动分割处理，支持 Checker 机制检查是否需要分割
* 内置三种 Checker，分别是 TimeChecker，SizeChecker，CountChecker
* 多种检查器可以叠加效果，只要其中有一个达到分割的条件都会进行分割

### v0.3.0-alpha

> 此版本发布于 2020-11-28

* 大幅度重构版本，废除了 handler 设计
* 暂不支持配置文件，正式版会支持
* v0.3.x 的第一个体验版

### v0.2.10

> 此版本发布于 2020-08-22

* 完善配置文件注释，目前是 // 和 # 都支持，后续将只支持 //
* 完善配置文件格式，目前的 Json 配置是不用 {} 包裹也可以的，后续将必须使用 {} 进行包裹

### v0.2.9

> 此版本发布于 2020-08-08

* 日志文件输出会自动创建父级不存在的目录

### v0.2.8

> 此版本发布于 2020-07-30

* 去掉 JsonEncoder 的 key 空格

doc.go
// Copyright 2022 FishGoddess. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
Package logit provides an easy way to use foundation for your logging operations.

1. basic:

	// Create a new logger for use.
	// Default level is debug, so all logs will be logged.
	// Invoke Close() isn't necessary in all situations.
	// If logger's writer has buffer or something like that, it's better to invoke Close() for syncing buffer or something else.
	logger := logit.NewLogger()
	//defer logger.Close()

	// Then, you can log anything you want.
	// Remember, logs will be ignored if their level is smaller than logger's level.
	// Log() will do some finishing work, so this invocation is necessary.
	logger.Debug("This is a debug message").Log()
	logger.Info("This is an info message").Log()
	logger.Warn("This is a warn message").Log()
	logger.Error("This is an error message").Log()
	logger.Error("This is a %s message, with format", "error").Log() // Format with params.

	// As you know, we provide some levels: debug, info, warn, error, off.
	// The lowest is debug and the highest is off.
	// If you want to change the level of your logger, do it at creating.
	logger = logit.NewLogger(logit.Options().WithWarnLevel())
	logger.Debug("This is a debug message, but ignored").Log()
	logger.Info("This is an info message, but ignored").Log()
	logger.Warn("This is a warn message, not ignored").Log()
	logger.Error("This is an error message, not ignored").Log()

	// Also, we provide some "old school" log method :)
	// (Don't mistake~ I love old school~)
	logger.Printf("This is a log %s, and it's for compatibility", "printed")
	logger.Print("This is a log printed, and it's for compatibility", 123)
	logger.Println("This is a log printed, and it's for compatibility", 666)

	// If you want to log with some fields, try this:
	user := struct {
		ID   int64  `json:"id"`
		Name string `json:"name"`
		Age  int    `json:"age"`
	}{
		ID:   666,
		Name: "FishGoddess",
		Age:  3,
	}
	logger.Warn("This is a structured message").Any("user", user).Json("userJson", user).Log()
	logger.Error("This is a structured message").Error("err", io.EOF).Int("trace", 123).Log()

	// You may notice logit.Options() which returns an options list.
	// Here is some of them:
	options := logit.Options()
	options.WithCaller()                          // Let logs carry caller information.
	options.WithLevelKey("lvl")                   // Change logger's level key to "lvl".
	options.WithWriter(os.Stderr)                 // Change logger's writer to os.Stderr without buffer or batch.
	options.WithBufferWriter(os.Stderr)           // Change logger's writer to os.Stderr with buffer.
	options.WithBatchWriter(os.Stderr)            // Change logger's writer to os.Stderr with batch.
	options.WithErrorWriter(os.Stderr)            // Change logger's error writer to os.Stderr without buffer or batch.
	options.WithTimeFormat("2006-01-02 15:04:05") // Change the format of time (Only the log's time will apply it).

	// You can bind context with logger and use it as long as you can get the context.
	ctx := logit.NewContext(context.Background(), logger)
	logger = logit.FromContext(ctx)
	logger.Info("Logger from context").Log()

	// You can initialize the global logger if you don't want to use an independent logger.
	// WithCallerDepth will set the depth of caller, and default is core.CallerDepth.
	// Functions in global logger are wrapped so depth of caller should be increased 1.
	// You can specify your depth if you wrap again or have something else reasons.
	logger = logit.NewLogger(options.WithCallerDepth(global.CallerDepth + 1))
	logit.SetGlobal(logger)
	logit.Info("Info from logit").Log()

	// We don't recommend you to call logit.SetGlobal unless you really need to call.
	// Instead, we recommend you to call logger.SetToGlobal to set one logger to global if you need.
	logger.SetToGlobal()
	logit.Println("Println from logit")

2. options:

	// We provide some options for you.
	options := logit.Options()
	options.WithDebugLevel()
	options.WithInfoLevel()
	options.WithWarnLevel()
	options.WithErrorLevel()
	options.WithAppender(appender.Text())
	options.WithDebugAppender(appender.Text())
	options.WithInfoAppender(appender.Text())
	options.WithWarnAppender(appender.Text())
	options.WithErrorAppender(appender.Text())
	options.WithPrintAppender(appender.Text())
	options.WithWriter(os.Stderr)
	options.WithBufferWriter(os.Stdout)
	options.WithBatchWriter(os.Stdout)
	options.WithDebugWriter(os.Stderr)
	options.WithInfoWriter(os.Stderr)
	options.WithWarnWriter(os.Stderr)
	options.WithErrorWriter(os.Stderr)
	options.WithPrintWriter(os.Stderr)
	options.WithPID()
	options.WithCaller()
	options.WithMsgKey("msg")
	options.WithTimeKey("time")
	options.WithLevelKey("level")
	options.WithPIDKey("pid")
	options.WithFileKey("file")
	options.WithLineKey("line")
	options.WithFuncKey("func")
	options.WithTimeFormat(appender.UnixTimeFormat) // UnixTimeFormat means time will be logged as unix time, an int64 number.
	options.WithCallerDepth(3)                      // Set caller depth to 3 so the log will get the third depth caller.
	options.WithInterceptors()

	// Remember, these options is only used for creating a logger.
	logger := logit.NewLogger(
		options.WithPID(),
		options.WithWriter(os.Stdout),
		options.WithTimeFormat("2006/01/02 15:04:05"),
		options.WithCaller(),
		options.WithCallerDepth(4),
		// ...
	)

	defer logger.Close()
	logger.Info("check options").Log()

	// You can use many options at the same time, but some of them is exclusive.
	// So only the last one in order will take effect if you use them at the same time.
	logit.NewLogger(
		options.WithDebugLevel(),
		options.WithInfoLevel(),
		options.WithWarnLevel(),
		options.WithErrorLevel(), // The level of logger is error.
	)

	// You can customize an option for your logger.
	// Actually, Option is just a function like func(logger *Logger).
	// So you can do what you want in creating a logger.
	syncOption := func(logger *logit.Logger) {
		go func() {
			select {
			case <-time.Tick(time.Second):
				logger.Sync()
			}
		}()
	}

	logit.NewLogger(syncOption)

3. appender:

	// We provide some ways to change the form of logs.
	// Actually, appender is an interface with some common methods, see appender.Appender.
	appender.Text()
	appender.Json()

	// Set appender to the one you want to use when creating a logger.
	// Default appender is appender.Text().
	logger := logit.NewLogger()
	logger.Info("appender.Text()").Log()

	// You can switch appender to the other one, such appender.Json().
	logger = logit.NewLogger(logit.Options().WithAppender(appender.Json()))
	logger.Info("appender.Json()").Log()

	// Every level has its own appender so you can append logs in different level with different appender.
	logger = logit.NewLogger(
		logit.Options().WithDebugAppender(appender.Text()),
		logit.Options().WithInfoAppender(appender.Text()),
		logit.Options().WithWarnAppender(appender.Json()),
		logit.Options().WithErrorAppender(appender.Json()),
	)

	// Appender is an interface so you can implement your own appender.
	// However, we don't recommend you to do that.
	// This interface may change in every version, so you will pay lots of extra attention to it.
	// So you should implement it only if you really need to do.

4. writer:

	// As you know, writer in logit is customized, not io.Writer.
	// The reason why we create a new Writer interface is we want a sync-able writer.
	// Then, we notice a sync-able writer also need a close method to sync all data in buffer when closing.
	// So, a new Writer is born:
	//
	//     type Writer interface {
	//	       Syncer
	//	       io.WriteCloser
	//     }
	//
	// In package writer, we provide some writers for you.
	writer.Wrap(os.Stdout)   // Wrap io.Writer to writer.Writer.
	writer.Buffer(os.Stderr) // Wrap io.Writer to writer.Writer with buffer, which needs invoking Sync() or Close().

	// Use the writer without buffer.
	logger := logit.NewLogger(logit.Options().WithWriter(os.Stdout))
	logger.Info("WriterWithoutBuffer").Log()

	// Use the writer with buffer, which is good for io.
	logger = logit.NewLogger(logit.Options().WithBufferWriter(os.Stdout))
	logger.Info("WriterWithBuffer").Log()
	logger.Sync() // Remember syncing data or syncing by Close().
	logger.Close()

	// Use the writer with batch, which is also good for io.
	logger = logit.NewLogger(logit.Options().WithBatchWriter(os.Stdout))
	logger.Info("WriterWithBatch").Log()
	logger.Sync() // Remember syncing data or syncing by Close().
	logger.Close()

	// Every level has its own appender so you can append logs in different level with different appender.
	logger = logit.NewLogger(
		logit.Options().WithBufferWriter(os.Stdout),
		logit.Options().WithBatchWriter(os.Stdout),
		logit.Options().WithWarnWriter(os.Stdout),
		logit.Options().WithErrorWriter(os.Stdout),
	)

	// Let me explain buffer writer and batch writer.
	// Both of them are base on a byte buffer and merge some writes to one write.
	// Buffer writer will write data in buffer to underlying writer if bytes in buffer are too much.
	// Batch writer will write data in buffer to underlying writer if writes to buffer are too much.
	//
	// Let's see something more interesting:
	// A buffer writer with buffer size 16 KB and a batch writer with batch count 64, whose performance is better?
	//
	// 1. Assume one log is 512 Bytes and its size is fixed
	// In buffer writer, it will merge 32 writes to 1 writes (16KB / 512Bytes);
	// In batch writer, it will always merge 64 writes to 1 writes;
	// Batch writer wins the game! Less writes means it's better to IO.
	//
	// 2. Assume one log is 128 Bytes and its size is fixed
	// In buffer writer, it will merge 128 writes to 1 writes (16KB / 128Bytes);
	// In batch writer, it will always merge 64 writes to 1 writes;
	// Buffer writer wins the game! Less writes means it's better to IO.
	//
	// 3. How about one log is 256 Bytes and its size is fixed
	// In buffer writer, it will merge 64 writes to 1 writes (16KB / 256Bytes);
	// In batch writer, it will always merge 64 writes to 1 writes;
	// They are the same in writing times.
	//
	// Based on what we mentioned above, we can tell the performance of buffer writer is depends on the size of log, and the batch writer is more stable.
	// Actually, the size of logs in production isn't fixed-size, so batch writer may be a better choice.
	// However, the buffer in batch writer is out of our control, so it may grow too large if our logs are too large.
	writer.Buffer(os.Stdout)
	writer.Batch(os.Stdout)
	writer.BufferWithSize(os.Stdout, 16*size.KB)
	writer.BatchWithCount(os.Stdout, 64)

5. global:

	// There are some global settings for optimizations, and you can set all of them in need.

	// 1. LogMallocSize (The pre-malloc size of a new Log data)
	// If your logs are extremely long, such as 4000 bytes, you can set it to 4096 to avoid re-malloc.
	global.LogMallocSize = 4 * size.MB

	// 2. WriterBufferSize (The default size of buffer writer)
	// If your logs are extremely long, such as 16 KB, you can set it to 2048 to avoid re-malloc.
	global.WriterBufferSize = 32 * size.KB

	// 3. MarshalToJson (The marshal function which marshal interface{} to json data)
	// Use std by default, and you can customize your marshal function.
	global.MarshalToJson = json.Marshal

	// After setting global settings, just use Logger as normal.
	logger := logit.NewLogger()
	defer logger.Close()

	logger.Info("set global settings").Uint64("LogMallocSize", global.LogMallocSize).Uint64("WriterBufferSize", global.WriterBufferSize).Log()

6. context:

	// By NewContext, you can bind a context with a logger and get it from context again.
	// So you can use this logger from everywhere as long as you can get this context.
	ctx := logit.NewContext(context.Background(), logit.NewLogger())

	// FromContext returns the logger in context.
	logger := logit.FromContext(ctx)
	logger.Info("This is a message logged by logger from context").Log()

	// Actually, you also have a chance to specify the key of logger in context.
	// It gives you a way to discriminate different businesses in using logger.
	// For example, you can create two loggers for your two different usages and
	// set them to a context with different key, so you can get each logger from context with each key.
	businessOneKey := "businessOne"
	logger = logit.NewLogger(logit.Options().WithMsgKey("businessOneMsg"))
	ctx = logit.NewContextWithKey(context.Background(), businessOneKey, logger)

	businessTwoKey := "businessTwo"
	logger = logit.NewLogger(logit.Options().WithMsgKey("businessTwoMsg"))
	ctx = logit.NewContextWithKey(ctx, businessTwoKey, logger)

	// Get different logger from the same context with different key.
	logger = logit.FromContextWithKey(ctx, businessOneKey)
	logger.Info("This is a message logged by logger from context with businessOneKey").Log()

	logger = logit.FromContextWithKey(ctx, businessTwoKey)
	logger.Info("This is a message logged by logger from context with businessTwoKey").Log()

7. caller:

	// Let's create a logger without caller information.
	logger := logit.NewLogger()
	logger.Info("I am without caller").Log()

	// We provide a way to add caller information to log even logger doesn't carry caller.
	logger.Info("Invoke log.WithCaller()").WithCaller().Log()
	logger.Close()

	time.Sleep(time.Second)

	// Now, let's create a logger with caller information.
	logger = logit.NewLogger(logit.Options().WithCaller())
	logger.Info("I am with caller").Log()

	// We won't carry caller information twice or more if logger carries caller information already.
	logger.Info("Invoke log.WithCaller() again").WithCaller().Log()
	logger.Close()

8. interceptor:

	// serverInterceptor is the global interceptor applied to all logs.
	func serverInterceptor(ctx context.Context, log *logit.Log) {
		log.String("server", "logit.interceptor")
	}

	// traceInterceptor is the global interceptor applied to all logs.
	func traceInterceptor(ctx context.Context, log *logit.Log) {
		trace, ok := ctx.Value("trace").(string)
		if !ok {
			trace = "unknown trace"
		}

		log.String("trace", trace)
	}

	// userInterceptor is the global interceptor applied to all logs.
	func userInterceptor(ctx context.Context, log *logit.Log) {
		user, ok := ctx.Value("user").(string)
		if !ok {
			user = "unknown user"
		}

		log.String("user", user)
	}

	// businessInterceptor is the log-level interceptor applied to one/some logs.
	func businessInterceptor(ctx context.Context, log *logit.Log) {
		business, ok := ctx.Value("business").(string)
		if !ok {
			business = "unknown business"
		}

		log.String("business", business)
	}

	// Use logit.Options().WithInterceptors to append some interceptors.
	logger := logit.NewLogger(logit.Options().WithInterceptors(serverInterceptor, traceInterceptor, userInterceptor))
	defer logger.Close()

	// By default, context passed to interceptor is context.Background().
	logger.Info("try interceptor - round one").Log()

	// You can use WithContext to change context passed to interceptor.
	ctx := context.WithValue(context.Background(), "trace", "666")
	ctx = context.WithValue(ctx, "user", "FishGoddess")
	logger.Info("try interceptor - round two").WithContext(ctx).Log()

	// The interceptors appended to logger will apply to all logs.
	// You can use Intercept to intercept one log rather than all logs.
	logger.Info("try interceptor - round three").WithContext(ctx).Intercept(businessInterceptor).Log()

	// Notice that WithContext should be called before Intercept if you want to pass this context to Intercept.
	ctx = context.WithValue(ctx, "business", "logger")
	logger.Info("try interceptor - round four").WithContext(ctx).Intercept(businessInterceptor).Log()

9. file:

	func mustCreateFile(filePath string) *os.File {
		f, err := os.OpenFile(filePath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
		if err != nil {
			panic(err)
		}

		return f
	}

	// Logger will log everything to console by default.
	logger := logit.NewLogger()
	logger.Info("I log everything to console.").Log()

	// You can use WithWriter to change writer in logger.
	logger = logit.NewLogger(logit.Options().WithWriter(os.Stdout))
	logger.Info("I also log everything to console.").Log()

	// As we know, we always log everything to file in production.
	logFile := filepath.Join(os.TempDir(), "test.log")
	fmt.Println(logFile)

	logger = logit.NewLogger(logit.Options().WithWriter(mustCreateFile(logFile)))
	logger.Info("I log everything to file.").String("logFile", logFile).Log()
	logger.Close()

	// Also, as you can see, there is a parameter called withBuffer in WithWriter option.
	// It will use a buffer writer to write logs if withBuffer is true which will bring a huge performance improvement.
	logFile = filepath.Join(os.TempDir(), "test_buffer.log")
	fmt.Println(logFile)

	logger = logit.NewLogger(logit.Options().WithWriter(mustCreateFile(logFile)))
	logger.Info("I log everything to file with buffer.").String("logFile", logFile).Log()
	logger.Close()

	// We provide some high-performance file for you. Try these:
	writer.BufferWithSize(os.Stdout, 128*size.KB)
	writer.BatchWithCount(os.Stdout, 256)
	logit.Options().WithBufferWriter(os.Stdout)
	logit.Options().WithBatchWriter(os.Stdout)

	// Wait a minute, we also provide a powerful file for you!
	// See extension/file/file.go.
	// It will rotate file and clean backups automatically.
	// You can set maxSize, maxAge and maxBackups by options.
	logFile = filepath.Join(os.TempDir(), "test_powerful.log")

	f, err := file.New(logFile)
	if err != nil {
		panic(err)
	}

	defer f.Close()

	_, err = f.Write([]byte("xxx"))
	if err != nil {
		panic(err)
	}

10. error:

	// uselessWriter is a demo writer to demonstrate the error handling function.
	type uselessWriter struct{}

	func (uw uselessWriter) Write(p []byte) (n int, err error) {
		return 0, errors.New("always error in writing")
	}

	// You can specify a function to handle errors happens in logger.
	// For example, you can count these errors and report them to team members by email.
	global.HandleError = func(name string, err error) {
		fmt.Printf("%s received an error: %+v\n", name, err)
	}

	// Let's log something to see what happen.
	logger := logit.NewLogger(logit.Options().WithWriter(&uselessWriter{}))
	logger.Info("See what happen?").Log()

11. config:

	// We provide a config which can be converted to option in logit.
	// It has many tags in fields, such json, yaml, toml, which means you can use config file to create logger.
	// You just need to define your config file then unmarshal your config file to this config.
	// Of course, you can embed this struct to your application config struct!
	cfg := config.Config{
		Level:         config.LevelDebug,
		TimeKey:       "x.time",
		LevelKey:      "x.level",
		MsgKey:        "x.msg",
		PIDKey:        "x.pid",
		FileKey:       "x.file",
		LineKey:       "x.line",
		FuncKey:       "x.func",
		TimeFormat:    config.UnixTimeFormat,
		WithPID:       true,
		WithCaller:    true,
		CallerDepth:   0,
		AutoSync:      "10s",
		Appender:      config.AppenderText,
		DebugAppender: config.AppenderText,
		InfoAppender:  config.AppenderText,
		WarnAppender:  config.AppenderText,
		ErrorAppender: config.AppenderText,
		PrintAppender: config.AppenderJson,
		Writer: config.WriterConfig{
			Target:     config.WriterTargetStdout,
			Mode:       config.WriterModeDirect,
			BufferSize: "4MB",
			BatchCount: 1024,
			Filename:   "test.log",
			DirMode:    0755,
			FileMode:   0644,
			TimeFormat: "20060102150405",
			MaxSize:    "128MB",
			MaxAge:     "30d",
			MaxBackups: 32,
		},
		DebugWriter: config.WriterConfig{},
		InfoWriter:  config.WriterConfig{},
		WarnWriter:  config.WriterConfig{},
		ErrorWriter: config.WriterConfig{},
		PrintWriter: config.WriterConfig{},
	}

	// Once you got a config, use Options() to convert to option in logger.
	opts, err := cfg.Options()
	if err != nil {
		panic(err)
	}

	fmt.Println(opts)

	// Then you can create your logger by options.
	// Amazing!
	logger := logit.NewLogger(opts...)
	defer logger.Close()

	logger.Info("My mother is a config").Any("config", cfg).Log()
	logger.Info("See logger").Any("logger", logger).Log()
*/
package logit // import "github.com/creky/logit"

const (
	// Version is the version string representation of logit.
	Version = "v0.5.5-alpha"
)
